x/staking/keeper/grpc_query.go

```diff
@@ -172,29 +172,39 @@ func (k Querier) ValidatorUnbondingDelegations(ctx context.Context, req *types.Q
 	if req.ValidatorAddr == "" {
 		return nil, status.Error(codes.InvalidArgument, "validator address cannot be empty")
 	}
-	var ubds types.UnbondingDelegations
 
 	valAddr, err := k.validatorAddressCodec.StringToBytes(req.ValidatorAddr)
 	if err != nil {
 		return nil, err
 	}
 
 	store := runtime.KVStoreAdapter(k.storeService.OpenKVStore(ctx))
-	srcValPrefix := types.GetUBDsByValIndexKey(valAddr)
-	ubdStore := prefix.NewStore(store, srcValPrefix)
-	pageRes, err := query.Paginate(ubdStore, req.Pagination, func(key, value []byte) error {
-		storeKey := types.GetUBDKeyFromValIndexKey(append(srcValPrefix, key...))
-		storeValue := store.Get(storeKey)
+	keys, pageRes, err := query.CollectionPaginate(
+		ctx,
+		k.UnbondingDelegationByValIndex,
+		req.Pagination,
+		func(key collections.Pair[sdk.ValAddress, sdk.AccAddress], value []byte) (collections.Pair[sdk.ValAddress, sdk.AccAddress], error) {
+			return key, nil
+		},
+		query.WithCollectionPaginationPairPrefix[sdk.ValAddress, sdk.AccAddress](valAddr),
+	)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// loop over the collected keys and fetch unbonding delegations
+	var ubds []types.UnbondingDelegation
+	for _, key := range keys {
```

Comment:
why can't we do this directly inside `query.CollectionPaginate`?

Reply:
see comment https://github.com/cosmos/cosmos-sdk/pull/17291#discussion_r1302707633